tests/TestCase/Database/ConnectionTest.php

```diff
@@ -1021,6 +1021,12 @@ public function testLoggerDecoratorDoesNotPrematurelyFetchRecords()
 
         $messages = Log::engine('queries')->read();
         $this->assertCount(1, $messages);
+
+        $st = $this->connection->execute('SELECT * FROM things WHERE id = 0');
+        $this->assertSame(0, $st->rowCount());
+
+        $messages = Log::engine('queries')->read();
+        $this->assertCount(2, $messages, 'Select queries without any matching rows should also be logged.');
```

Comment:
The problem with 0 results was in particular for buffered queries that do create result sets, that is queries made with the query builder. I think it would be good to have an additional test case for that?

Reply:
I added additional test case. 

For buffered queries without any matching rows `ResultSet` creates an `SplFixedFixed` of size `0`. Hence when you try to iterate it a `StatementInterface::fetch()` is never executed, which is why the query was never logged.

https://github.com/cakephp/cakephp/blob/edf991c10bd365981217d436b08d5612f538884d/src/ORM/ResultSet.php#L175-L178

Thus moving the logging to `LoggingStatment::rowCount()` (which is called by `ResultSet::count()`) fixes the issue.